src/utilities/bmi/mass_balance.cpp

```diff
@@ -0,0 +1,194 @@
+                INPUT_MASS_NAME << "(" << mass_in <<  ") - " <<
+                OUTPUT_MASS_NAME << " (" << mass_out <<  ") - " << 
+                STORED_MASS_NAME << " (" << mass_stored <<  ") - " << 
+                LEAKED_MASS_NAME << " (" << mass_leaked << ") = " << 
+                mass_balance << "\n\t" << "tolerance: " << tolerance << std::endl;
+            return make_unexpected<ProtocolError>( ProtocolError(
+                is_fatal ? Error::PROTOCOL_ERROR : Error::PROTOCOL_WARNING,
+                ss.str()
+                )
+            );
+        }
+
+    }
+    return {};
+}
+
+auto NgenMassBalance::check_support(const ModelPtr& model) -> expected<void, ProtocolError> {
+    if (model != nullptr && model->is_model_initialized()) {
+        double mass_var;
+        std::vector<std::string> units;
+        units.reserve(4);
+        try{
+            for(const auto& name : 
+                {INPUT_MASS_NAME, OUTPUT_MASS_NAME, STORED_MASS_NAME, LEAKED_MASS_NAME}
+            ){
+                model->GetValue(name, &mass_var);
+                units.push_back( model->GetVarUnits(name) );
+            }
+            //Compare all other units to the first one (+1)
+            if( std::equal( units.begin()+1, units.end(), units.begin() ) ) {
+                this->supported = true;
+                return {};
+            }
+            else{
+                // It may be possible to do unit conversion and still do meaninful mass balance
+                // this could be added as an extended feature, but for now, I don't think this is
+                // worth the complexity.  It is, however, worth the sanity check performed here
+                // to ensure the units are consistent.
+                return make_unexpected<ProtocolError>( ProtocolError(
+                    Error::INTEGRATION_ERROR,
+                    "mass_balance: variables have incosistent units, cannot perform mass balance."
+                    )
+                );
+            }
+        } catch (const std::exception &e) {
+            std::stringstream ss;
+            ss << "mass_balance: Error getting mass balance values for module '" << model->GetComponentName() << "': " << e.what() << std::endl;
+            return make_unexpected<ProtocolError>( ProtocolError(
+                Error::INTEGRATION_ERROR,
+                ss.str()
+                )
+            );
+        }
+    } else {
+        return make_unexpected<ProtocolError>( ProtocolError(
+            Error::UNITIALIZED_MODEL,
+            "Cannot check mass balance for uninitialized model. Disabling mass balance protocol."
+            )
+        );
+    }
+    return {};
+}
+
+auto NgenMassBalance::initialize(const ModelPtr& model, const Properties& properties) -> expected<void, ProtocolError>
+{
+    //Ensure the model is capable of mass balance using the protocol
+    check_support(model).or_else( error_or_warning );
```

Comment:
Coming back to this, I think we should check model support only after reading the configuration. Otherwise, this will raise an exception for models that don't support the mass balance checker _and_ aren't configured.